collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java

```diff
@@ -0,0 +1,708 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.zip.GZIPInputStream;
+
+import timber.log.Timber;
+
+public class HttpClientConnection implements OpenRosaHttpInterface {
+
+    private static final String USER_AGENT_HEADER = "User-Agent";
+    private static final String OPEN_ROSA_VERSION_HEADER = "X-OpenRosa-Version";
+    private static final String OPEN_ROSA_VERSION = "1.0";
+    private static final String DATE_HEADER = "Date";
+    private static final String ACCEPT_ENCODING_HEADER = "Accept-Encoding";
+    private static final String GZIP_CONTENT_ENCODING = "gzip";
+
+    private static final int CONNECTION_TIMEOUT = 30000;
+    private static final int UPLOAD_CONNECTION_TIMEOUT = 60000; // it can take up to 27 seconds to spin up an Aggregate
+    private static final String HTTP_CONTENT_TYPE_TEXT_XML = "text/xml";
+
+    private CredentialsProvider credentialsProvider;
+    private CookieStore cookieStore;
+
+    public HttpClientConnection() {
+        credentialsProvider = new AgingCredentialsProvider(7 * 60 * 1000);
+        cookieStore = new BasicCookieStore();
+    }
+
+    private enum ContentTypeMapping {
+        XML("xml",  ContentType.TEXT_XML),
+        _3GPP("3gpp", ContentType.create("audio/3gpp")),
+        _3GP("3gp",  ContentType.create("video/3gpp")),
+        AVI("avi",  ContentType.create("video/avi")),
+        AMR("amr",  ContentType.create("audio/amr")),
+        CSV("csv",  ContentType.create("text/csv")),
+        JPG("jpg",  ContentType.create("image/jpeg")),
+        MP3("mp3",  ContentType.create("audio/mp3")),
+        MP4("mp4",  ContentType.create("video/mp4")),
+        OGA("oga",  ContentType.create("audio/ogg")),
+        OGG("ogg",  ContentType.create("audio/ogg")),
+        OGV("ogv",  ContentType.create("video/ogg")),
+        WAV("wav",  ContentType.create("audio/wav")),
+        WEBM("webm", ContentType.create("video/webm")),
+        XLS("xls",  ContentType.create("application/vnd.ms-excel"));
+
+        private String extension;
+        private ContentType contentType;
+
+        ContentTypeMapping(String extension, ContentType contentType) {
+            this.extension = extension;
+            this.contentType = contentType;
+        }
+
+        public static ContentType of(String fileName) {
+            String extension = FileUtils.getFileExtension(fileName);
+
+            for (ContentTypeMapping m : values()) {
+                if (m.extension.equals(extension)) {
+                    return m.contentType;
+                }
+            }
+
+            return null;
+        }
+    }
+
+
+    @Override
+    public @NonNull
+    HttpGetResult get(@NonNull URI uri, @Nullable final String contentType, @Nullable HttpCredentialsInterface credentials) throws Exception {
+        addCredentialsForHost(uri, credentials);
+        getCookieStore().clear();
+
+        HttpContext localContext = getHttpContext();
+        HttpClient httpclient = createHttpClient(CONNECTION_TIMEOUT);
```

Comment:
Would be a good practice to inject `HttpClient` creation as well.

Reply:
Similar response as above, the intention is to swap out HttpClientConnection, but also it is supposed to be the only class that knows anything about httpclient.